[feat] Utilize React 18 createRoot API

sfc-gh-bnisco opened the pull request:
## Describe your changes

- This brings Streamlit onto the React 18 `createRoot` API.
- This PR is a long-running feature branch. Every individual commit into this branch has been independently reviewed as separate PRs. You can click through on each commit's attached PR to see them.
- In general, the changes in each PR fall under one of the follow cases:
  1. `createRoot` setup/infra changes
  2. Source changes to fix things like timing or set state issues due to `createRoot`
  3. Test updates to stabilize tests under `createRoot` mode

## GitHub Issue Link (if applicable)

## Testing Plan

- ✅ All unit tests pass
  - We did not skip any tests
- ✅ All e2e tests pass
  - We only skipped 1 test completely, and 4 tests in specific browsers. These 5 skipped e2e tests deal with dataframe
- ✅ Manually tested scenarios with the [mega tester app](https://create-root-demo.streamlit.app/) using this wheel file

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-bnisco committed:
[feat] Utilize React 18 createRoot API (#10399)

Note: This is a PR into the long-running feature branch
`feature/react-create-root`, not `develop`! Failing tests are expected
while we work through the long tail of issues.

- This implements Phase 2 of the React 18 `createRoot` plan by
introducing just the minimally required bootstrapping logic to utilize
the new API.

- Updates the test utils to ensure they are running in React 18 mode

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-bnisco committed:
[fix] Fixes JS unit tests in createRoot mode (#10435)

## Describe your changes

Note: This is a PR into the long-running feature branch
`feature/react-create-root`, not `develop`! Failing tests are expected
while we work through the long tail of issues.

- Fixes all JS unit tests while running in `createReact` mode. There
were previously a number of them failing.

## GitHub Issue Link (if applicable)

## Testing Plan

- ✅ Updates unit tests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-bnisco committed:
[fix] Tooltip positioning logic with createRoot (#10436)

## Describe your changes

Note: This is a PR into the long-running feature branch
`feature/react-create-root`, not `develop`! Failing tests are expected
while we work through the long tail of issues.

- Fixes an issue where the Tooltip would appear visually incorrect when
rendering in `createRoot`
- Due to the new batching rendering logic, there was a timing issue when
attempting to measure position of elements. To fix this, we keep some
local state for the Tooltip's state and defer its measurement into a
`useEffect`

## GitHub Issue Link (if applicable)

## Testing Plan

- ✅ This fixes e2e tests that were previously failing for Tooltip

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-bnisco committed:
[fix] File uploader state management with createRoot (#10439)

## Describe your changes

Note: This is a PR into the long-running feature branch
`feature/react-create-root`, not `develop`! Failing tests are expected
while we work through the long tail of issues.

- Fixes state race conditions when using file uploader in `createRoot`
API
  - Leverages `flushState` since we are dealing with browser APIs
  - Removes a piece of unused state + unused function
- Result: all file uploader tests in CI now pass!

## GitHub Issue Link (if applicable)

## Testing Plan

- ✅ This fixes e2e tests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-bnisco committed:
[fix] Websocket connection state with createRoot (#10449)

## Describe your changes

Note: This is a PR into the long-running feature branch
`feature/react-create-root`, not `develop`! Failing tests are expected
while we work through the long tail of issues.

- Fixes an issue where connectionState updates must all applied in an
expected order. `createRoot` would sometimes batch these changes,
leading to "skipped" states
- Result: All websocket e2e tests are now passing

## GitHub Issue Link (if applicable)

## Testing Plan

- ✅ This fixes e2e tests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-bnisco committed:
[fix] Chat Input measurement with createRoot (#10474)

## Describe your changes

Note: This is a PR into the long-running feature branch
`feature/react-create-root`, not `develop`! Failing tests are expected
while we work through the long tail of issues.

- Adheres to React best practices for dealing with measurements by
leveraging `useLayoutEffect` for places in ChatInput that deal with
measurements
- Impact: Fixes the e2e snapshot tests for ChatInput that were failing
with `createRoot`

## GitHub Issue Link (if applicable)

## Testing Plan

- ✅ Stabilizes e2e tests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-bnisco committed:
[fix] Custom Component integration with createRoot (#10468)

## Describe your changes

Note: This is a PR into the long-running feature branch
`feature/react-create-root`, not `develop`! Failing tests are expected
while we work through the long tail of issues.

- Fixes state issues with custom components
- Impact: The Component-template Repo E2E Test suite now passes

## GitHub Issue Link (if applicable)

## Testing Plan

- ✅ Covered by the component-e2e suite

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-bnisco committed:
[fix] Run all RTL hook tests in createRoot mode (#10482)

Note: This is a PR into the long-running feature branch
`feature/react-create-root`, not `develop`! Failing tests are expected
while we work through the long tail of issues.

- Ensures that anything that uses [React Hooks Testing
Library](https://github.com/testing-library/react-hooks-testing-library)
leverages the new import import path so that they are run in
`createRoot` mode
- Note that this is following [this part in their
docs](https://github.com/testing-library/react-hooks-testing-library?tab=readme-ov-file#a-note-about-react-18-support).
- Impact: Removes any `ReactDOM.render is no longer supported in React
18.` logs in the test output by ensuring we are actually running the
hooks in `createRoot` mode!

- ✅ This is a test infrastructure change

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-bnisco committed:
[fix] Stabilize Arrow Vega Lite Chart with createRoot (#10517)

## Describe your changes

Note: This is a PR into the long-running feature branch
`feature/react-create-root`, not `develop`! Failing tests are expected
while we work through the long tail of issues.

- Utilizes `useLayoutEffect` in `ArrowVegaLiteChart` to prevent
unnecessary layout shifting.

## GitHub Issue Link (if applicable)

## Testing Plan

- ✅ Covered by existing e2e snapshot tests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-bnisco committed:
[fix] Dataframe test fixes with createRoot (#10483)

## Describe your changes

Note: This is a PR into the long-running feature branch
feature/react-create-root, not develop!

- Stabilizes as many dataframe-related tests as possible
- The biggest difference with dataframe and other tests is that Glide
uses a `canvas` under the hood. Given the changes with `createRoot`, the
timing of application of these changes caused many of the dataframe
tests (mostly those that deal with interaction) to be flaky.
- This PR includes a number of new test utils that stabilize the tests
by doing more intelligent waiting.
- As a last resort, some tests had to be skipped. Reasons for skipping
are written in comments in the code.
- This was the last big chunk of failing tests with `createRoot`, hence
why we actually have a clean build now!

## GitHub Issue Link (if applicable)

## Testing Plan

- ✅ This updates and stabilizes many e2e tests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-bnisco committed:
[fix] Update tooltip positioning logic + test fixes with createRoot (#10565)

## Describe your changes

Note: this is a PR into the `feature/react-create-root` feature branch!

### End-to-End Test Improvements:

*
[`e2e_playwright/multipage_apps/mpa_basics_test.py`](diffhunk://#diff-24ecc937a117bfd88cfd3a8cbfe878c86efb5b42eea1753af7c42405908574dfR61):
Added visibility checks for page headings in various tests to ensure the
correct page is displayed before proceeding with further actions.
[[1]](diffhunk://#diff-24ecc937a117bfd88cfd3a8cbfe878c86efb5b42eea1753af7c42405908574dfR61)
[[2]](diffhunk://#diff-24ecc937a117bfd88cfd3a8cbfe878c86efb5b42eea1753af7c42405908574dfR216-R217)
[[3]](diffhunk://#diff-24ecc937a117bfd88cfd3a8cbfe878c86efb5b42eea1753af7c42405908574dfR227-R228)
[[4]](diffhunk://#diff-24ecc937a117bfd88cfd3a8cbfe878c86efb5b42eea1753af7c42405908574dfR238)
[[5]](diffhunk://#diff-24ecc937a117bfd88cfd3a8cbfe878c86efb5b42eea1753af7c42405908574dfR279-R280)
[[6]](diffhunk://#diff-24ecc937a117bfd88cfd3a8cbfe878c86efb5b42eea1753af7c42405908574dfR308-R309)
[[7]](diffhunk://#diff-24ecc937a117bfd88cfd3a8cbfe878c86efb5b42eea1753af7c42405908574dfR337-R338)
*
[`e2e_playwright/st_chat_input_test.py`](diffhunk://#diff-91c04b9fe23d134b720135a67bd5b3ecc30f6af9629a7a5c4ece1de63ef19777L33-R35):
Improved the file upload helper by adding a keyboard escape press and
forcing a click to remove hover focus.
[[1]](diffhunk://#diff-91c04b9fe23d134b720135a67bd5b3ecc30f6af9629a7a5c4ece1de63ef19777L33-R35)
[[2]](diffhunk://#diff-91c04b9fe23d134b720135a67bd5b3ecc30f6af9629a7a5c4ece1de63ef19777L342-R344)
[[3]](diffhunk://#diff-91c04b9fe23d134b720135a67bd5b3ecc30f6af9629a7a5c4ece1de63ef19777R355-R360)

### Tooltip Side Effect Refactor:

*
[`frontend/lib/src/components/shared/Tooltip/Tooltip.tsx`](diffhunk://#diff-9c1459f9e01a40718ca2da673c5c1caab54c1be3e91c17f597d3bb2af657cd10L22):
Moved the inline tooltip positioning logic into its own
`useTooltipMeasurementSideEffect` hook.
[[1]](diffhunk://#diff-9c1459f9e01a40718ca2da673c5c1caab54c1be3e91c17f597d3bb2af657cd10L22)
[[2]](diffhunk://#diff-9c1459f9e01a40718ca2da673c5c1caab54c1be3e91c17f597d3bb2af657cd10R36)
[[3]](diffhunk://#diff-9c1459f9e01a40718ca2da673c5c1caab54c1be3e91c17f597d3bb2af657cd10L89-R89)
*
[`frontend/lib/src/components/shared/Tooltip/useTooltipMeasurementSideEffect.ts`](diffhunk://#diff-053a09a573084c04b7682e234a28bc26af583bfdc4692d7c92fdfd17a1601c9bR1-R101):
This is the standalone hook to handle tooltip positioning and
measurement with retry logic to ensure proper positioning even when
initial measurements are invalid.
*
[`frontend/lib/src/components/shared/Tooltip/useTooltipMeasurementSideEffect.test.tsx`](diffhunk://#diff-28497bcde1594bcc2c520240f4699c59cb3b8ebdaaa131d8ef3c9133d166c6c5R1-R223):
Added tests for the new `useTooltipMeasurementSideEffect` hook to verify
its behavior in various scenarios, including handling right and left
overflow and retrying measurements.

## GitHub Issue Link (if applicable)

## Testing Plan

- ✅ Covered by existing e2e test
- ✅ Adds unit tests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-nbellante commented on e2e_playwright/help_tooltip_test.py:
```diff
@@ -28,9 +30,10 @@ def test_tooltip_does_not_overflow_on_the_left_side(app: Page):
-    assert bounding_box["x"] >= 0
+    # Wait until the tooltip is positioned correctly
+    wait_until(
+        app, lambda: (bbox := tooltip.bounding_box()) is not None and bbox["x"] >= 0
```
lol TIL i learned this operator is referred to as the walrus operator. 🙌 name. no notes.

sfc-gh-nbellante reviewed (approved):
lgtm

sfc-gh-bnisco committed:
[refactor] Remove flushSync from deckgl (#10640)

## Describe your changes

- Removes `flushSync` from deck.gl codepaths by changing the rendering
order such that we only render the deck.gl component once we actually
have a `viewState` to render.
- This removes the timing issue that was previously requiring the
`flushSync`

## GitHub Issue Link (if applicable)

## Testing Plan

- ✅ existing e2e tests cover this
- ✅ manually tested

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-bnisco committed:
[fix] Remove a lifecycle warning with createRoot (#10641)

## Describe your changes

- Removes a console warning stemming from a usage of `flushSync` in
`App.tsx`

## GitHub Issue Link (if applicable)

## Testing Plan

- ✅ E2E tests cover the functionality of this

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.
